app/gen-server/lib/DocWorkerMap.ts

```diff
@@ -306,6 +310,38 @@ export class DocWorkerMap implements IDocWorkerMap {
     }
   }
 
+  public onWorkerUnavailable(workerInfo: DocWorkerInfo, cb: () => void): () => void {
+    let timeout: NodeJS.Timeout;
+    const group = workerInfo.group || 'default';
+    const workerId = workerInfo.id;
+
+    (function recursiveTimeout(self, nbFailures = 0) {
+      timeout = setTimeout(async () => {
+        if (nbFailures >= 3) {
+          log.error(
+            'DocWorkerMap: Presence checker failed 3 times, considering the worker "%s" is not available anymore',
+            workerId
+          );
+          return cb();
+        }
```

Comment:
I wonder whether this is useful or whether that's what we would want:
 - The only case I think we could ultimately enter in this condition is because of a failure when calling `sismemberAsync()`;
 - I tested simulating network issues (in my case, I simulated some `ECONNREFUSED`), no exception has been raised, and I never entered in this condition;
 - I feel like it is not the role of this method to handle connection issues with Redis, I see that errors are already logged [above in this class](https://github.com/gristlabs/grist-core/blob/af38519ce39cec8dc4565095e5e30d9168043893/app/gen-server/lib/DocWorkerMap.ts#L193);

I tend to think that we may just simplify everything:
 - not count the number of failure in this recursive setTimeout;
 - just log for each exception;
 - If we want to shutdown the worker because it cannot join the redis server, it's another subject so another issue/PR (that I can take care of);

What do you think?

Reply:
FWIW, I chose to do as I planed above.